extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,270 @@
+  }
+
+  @Override
+  public boolean start()
+  {
+    final Lock writeLock = startStopSync.writeLock();
+    writeLock.lock();
+    try {
+      if (!started.compareAndSet(false, true)) {
+        LOG.warn("Already started! [%s]", extractorID);
+        return true;
+      }
+      if (!manager.scheduleAndWait(extractorID, extractionNamespace, firstCacheTimeout)) {
+        LOG.warn("Failed to schedule lookup [%s]", extractorID);
+        return false;
+      }
+      LOG.debug("NamespaceLookupExtractorFactory[%s] started", extractorID);
+      return true;
+    }
+    finally {
+      writeLock.unlock();
+    }
+  }
+
+  @Override
+  public boolean close()
+  {
+    final Lock writeLock = startStopSync.writeLock();
+    writeLock.lock();
+    try {
+      if (!started.compareAndSet(true, false)) {
+        LOG.warn("Not started! [%s]", extractorID);
+        return true;
+      }
+      return manager.checkedDelete(extractorID);
+    }
+    finally {
+      writeLock.unlock();
+    }
+  }
+
+  @Override
+  public boolean replaces(@Nullable LookupExtractorFactory other)
+  {
+    if (other != null && other instanceof NamespaceLookupExtractorFactory) {
+      NamespaceLookupExtractorFactory that = (NamespaceLookupExtractorFactory) other;
+      if (isOneToOne() != ((NamespaceLookupExtractorFactory) other).isOneToOne()) {
+        return true;
+      }
+      if (getFirstCacheTimeout() != ((NamespaceLookupExtractorFactory) other).getFirstCacheTimeout()) {
+        return true;
+      }
+      return !extractionNamespace.equals(that.extractionNamespace);
+    }
+    return true;
+  }
+
+  @Override
+  public LookupIntrospectHandler getIntrospectHandler()
+  {
+    return lookupIntrospectHandler;
+  }
+
+  @JsonProperty
+  public ExtractionNamespace getExtractionNamespace()
+  {
+    return extractionNamespace;
+  }
+
+  @JsonProperty
+  public long getFirstCacheTimeout()
+  {
+    return firstCacheTimeout;
+  }
+
+  @JsonProperty
+  public boolean isOneToOne()
+  {
+    return oneToOne;
+  }
+
+  private String buildID()
+  {
+    return UUID.randomUUID().toString();
+  }
+
+  // Grab the latest snapshot from the cache manager
+  @Override
+  public LookupExtractor get()
+  {
+    final Lock readLock = startStopSync.readLock();
+    readLock.lock();
+    try {
+      if (!started.get()) {
+        throw new ISE("Factory [%s] not started", extractorID);
+      }
+      String preVersion = null, postVersion = null;
+      Map<String, String> map = null;
+      // Make sure we absolutely know what version of map we grabbed (for caching purposes)
+      do {
+        preVersion = manager.getVersion(extractorID);
+        if (preVersion == null) {
+          throw new ISE("Namespace vanished for [%s]", extractorID);
+        }
+        map = manager.getCacheMap(extractorID);
+        postVersion = manager.getVersion(extractorID);
+        if (postVersion == null) {
+          // We lost some horrible race... make sure we clean up
```

Comment:
No, read/write lock is only for start/stop races. This is to try and help prevent a race in delete.